File: libraries/dotnet/UrcConverter/Parser/Lex.fs
namespace UrcConverter.Parser

open System
open System.Globalization
open FsToolkit.ErrorHandling
open UrcConverter

module internal Lex =


    let intToken (token: string) line : Result<int, UrcError> =
        match Int32.TryParse(token, NumberStyles.Integer, CultureInfo.InvariantCulture) with
        | true, value -> Ok value
        | false, _ -> Error(UrcError.Syntax(line, $"invalid integer: '{token}'"))

    let floatToken (token: string) line : Result<float, UrcError> =
        match Double.TryParse(token, NumberStyles.Float, CultureInfo.InvariantCulture) with
        | true, value -> Ok value
        | false, _ -> Error(UrcError.Syntax(line, $"invalid float: '{token}'"))

    let meterToken (token: string) line : Result<Meter, UrcError> =
        match token.Split('/') with
        | [| beatsStr; noteValueStr |] ->
            match Int32.TryParse(beatsStr, NumberStyles.None, CultureInfo.InvariantCulture),
                  Int32.TryParse(noteValueStr, NumberStyles.None, CultureInfo.InvariantCulture) with
            | (true, beats), (true, noteValue) when beats >= 1 && noteValue >= 1 ->
                Ok { Beats = beats; NoteValue = noteValue }
            | _ -> Error(UrcError.Rule(17, line, $"invalid meter: '{token}'"))
        | _ -> Error(UrcError.Rule(17, line, $"invalid meter: '{token}'"))

    let layoutTypeToken (token: string) line : Result<int * int option, UrcError> =
        match token.Split('+') with
        | [| keysStr; specialStr |] ->
            match Int32.TryParse(keysStr, NumberStyles.None, CultureInfo.InvariantCulture),
                  Int32.TryParse(specialStr, NumberStyles.None, CultureInfo.InvariantCulture) with
            | (true, keys), (true, special) -> Ok(keys, Some special)
            | _ -> Error(UrcError.Syntax(line, $"invalid Type value: '{token}'"))
        | [| keysStr |] ->
            match Int32.TryParse(keysStr, NumberStyles.None, CultureInfo.InvariantCulture) with
            | true, keys -> Ok(keys, None)
            | false, _ -> Error(UrcError.Syntax(line, $"invalid Type value: '{token}'"))
        | _ -> Error(UrcError.Syntax(line, $"invalid Type value: '{token}'"))

    let private commaList emptyMessage parse (value: string) line =
        value.Split(',')
        |> Array.toList
        |> List.traverseResultM (fun token ->
            let token = token.Trim()

            if token = "" then
                Result.Error(UrcError.Syntax(line, emptyMessage))
            else
                parse token line)

    let floatList value line =
        commaList "empty value in list" floatToken value line

    let intList emptyMessage value line =
        commaList emptyMessage intToken value line

File: libraries/python/src/urc_converter/parser/lex.py
"""Scalar and list token parsing for URC field values."""

import re

from ..error import UrcError
from ..model import Meter
from ..strings import SYNTAX, rule

_INT_RE = re.compile(r"^-?\d+$")
_FLOAT_RE = re.compile(r"^-?\d+(\.\d+)?$")
_METER_RE = re.compile(r"^(\d+)/(\d+)$")
_TYPE_RE = re.compile(r"^(\d+)(?:\+(\d+))?$")


def int_value(token: str, line: int) -> int:
	if _INT_RE.match(token) is None:
		raise UrcError(SYNTAX, line, f"invalid integer: {token!r}")

	return int(token)


def float_value(token: str, line: int) -> float:
	if _FLOAT_RE.match(token) is None:
		raise UrcError(SYNTAX, line, f"invalid float: {token!r}")

	return float(token)


def float_list(value: str, line: int) -> list[float]:
	values: list[float] = []

	for token in value.split(","):
		token = token.strip()
		if token == "":
			raise UrcError(SYNTAX, line, "empty value in list")

		values.append(float_value(token, line))

	return values


def layout_type(value: str, line: int) -> tuple[int, int]:
	match = _TYPE_RE.match(value)
	if match is None:
		raise UrcError(SYNTAX, line, f"invalid Type value: {value!r}")

	keys = int(match.group(1))
	special = int(match.group(2) or "0")
	if keys < 1 or (match.group(2) is not None and special < 1):
		raise UrcError(SYNTAX, line, "Type values must be positive")

	return keys, special


def meter(token: str, line: int) -> Meter:
	match = _METER_RE.match(token)
	if match is None:
		raise UrcError(rule(17), line, f"invalid meter: {token!r}")

	beats = int(match.group(1))
	note_value = int(match.group(2))
	if beats < 1 or note_value < 1:
		raise UrcError(rule(17), line, f"invalid meter: {token!r}")

	return Meter(beats, note_value)
